sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala

```diff
@@ -138,9 +138,8 @@ trait HashJoin extends JoinCodegenSupport {
     UnsafeProjection.create(streamedBoundKeys)
 
   @transient protected[this] lazy val boundCondition = if (condition.isDefined) {
-    if (joinType == FullOuter && buildSide == BuildLeft) {
-      // Put join left side before right side. This is to be consistent with
-      // `ShuffledHashJoinExec.fullOuterJoin`.
+    if ((joinType == FullOuter || joinType == LeftOuter) && buildSide == BuildLeft) {
+      // Put join left side before right side. This is to be consistent with ShuffledHashJoinExec.
```

Comment:
I wonder whether this whole `boundCondition` block can be simplified to

```
Predicate.create(condition.get, left.output ++ right.output).eval _
```

since the left side is always on the left and the right side is always on the right.

Further down in this file, we have

https://github.com/apache/spark/blob/aa81f424aee2db9d85f1ab6720a519b9552cc7a7/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala#L104-L107

showing that `buildPlan` and `streamPlan` are just re-mappings of `left` and `right` depending on the build side.

----

Several years ago, it looks like we used to do something similar to my proposal:

https://github.com/apache/spark/blob/0340b3d279de6be4903673bbf3e6a1a2653de6c0/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala#L119-L123

but we switched to the current use of `buildPlan` and `streamPlan` in a refactoring in https://github.com/apache/spark/pull/12102 (I'm not fully clear on why).

Reply:
+1, It's fragile to keep 2 pieces of code in sync.

Reply:
Oh I guess it's wrong for inner join + build left?

Reply:
I found this
```
  protected def createResultProjection(): (InternalRow) => InternalRow = joinType match {
    case LeftExistence(_) =>
      UnsafeProjection.create(output, output)
    case _ =>
      // Always put the stream side on left to simplify implementation
      // both of left and right side could be null
      UnsafeProjection.create(
        output, (streamedPlan.output ++ buildPlan.output).map(_.withNullability(true)))
  }
```

I think stream side left and build side right is the common case, but we have some special cases for outer joins.

Reply:
Actually I don't know why outer join needs to be a special case. Can't we always put streaming side left and build side right?

Reply:
HashJoin.scala
```scala
override def output: Seq[Attribute] = {
    joinType match {
      case _: InnerLike =>
        left.output ++ right.output
      case LeftOuter =>
        left.output ++ right.output.map(_.withNullability(true))
      case RightOuter =>
        left.output.map(_.withNullability(true)) ++ right.output
      case j: ExistenceJoin =>
        left.output :+ j.exists
      case LeftExistence(_) =>
        left.output
      case x =>
        throw new IllegalArgumentException(s"HashJoin should not take $x as the JoinType")
    }
  }
```


ShuffledJoin.scala
```scala
override def output: Seq[Attribute] = {
    joinType match {
      case _: InnerLike =>
        left.output ++ right.output
      case LeftOuter =>
        left.output ++ right.output.map(_.withNullability(true))
      case RightOuter =>
        left.output.map(_.withNullability(true)) ++ right.output
      case FullOuter =>
        (left.output ++ right.output).map(_.withNullability(true))
      case j: ExistenceJoin =>
        left.output :+ j.exists
      case LeftExistence(_) =>
        left.output
      case x =>
        throw new IllegalArgumentException(
          s"${getClass.getSimpleName} not take $x as the JoinType")
    }
  }
```


@cloud-fan  Output determines that we cannot simply put streaming side left and build side right.  
Am I understanding correctly?

Reply:
Yea that's why we have a `createResultProjection` to reorder the columns.

Reply:
Ah, I think I understand now:

In all of the `*Join` methods in this file (which _doesn't_ include full outer join), the streamed side is always the left side of the `joinRow` that is passed to `boundCondition`: in `HashJoin.scala`, the input to `boundCondition` is not necessarily the same as the output of the join operator itself (for that, we have `resultProj` which comes from `createResultProjection` which remains in sync with `boundCondition` ([source](https://github.com/apache/spark/blob/aa81f424aee2db9d85f1ab6720a519b9552cc7a7/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala#L151-L159)).

However, `boundCondition` is inherited in `ShuffledHashJoinExec` and there we have a `buildSideOrFullOuterJoin` method ([source](https://github.com/apache/spark/blob/aa81f424aee2db9d85f1ab6720a519b9552cc7a7/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledHashJoinExec.scala#L119-L158)) and there the `resultProjection` operates over the same input schema as the join output schema (rather than assuming that the joined row always has the streamed side on the left). In that code, `boundCondition` is evaluated over an input that matches the join output schema and there the streamed side could be on either side rather than only the left.

I think this inheritance is confusing and hard to reason about. 

It seems like `HashJoin.scala` has an invariant of "streamed side always on left" which gets violated in `ShuffleHashJoin.scala`'s separate implementation of outer joins.

I wonder whether we can address this bug by modifying `ShuffledHashJoinExec.buildSideOrFullOuterJoin ` so that it always unconditionally places the streamed side on the left (the same as `HashJoin.scala`'s default).